### PR TITLE
Fix bug 1688570: improve performance of initial editor change

### PR DESCRIPTION
**Pull request by julen**
Changing from/to initial textarea values would result in redundant
re-renders due to subscribing to `unsavedChangesExist` and
`unsavedChangesIgnored` redux state changes.

Implementation-wise, this accesses `unsavedchanges` state directly,
without subscribing to its changes because the state members are not
used for rendering purposes.

For direct store access react-redux strongly suggests to leverage the
`useStore()` API, because accessing the store via context is an
implementation detail and it might change in the future. That's why the
components connecting redux state -> props have been converted to hooks.

It'd also have been possible to avoid these intermediate components and
co-locate redux state directly within components, however that would
require more invasive changes, creeping the scope of this change
considerably.

Refs. https://react-redux.js.org/using-react-redux/accessing-store

----

Before:
before

After:
after

### CODE REVIEW

**Comment by julen** [comment]
Tests for changed components are currently broken. I've been banging my head to find a solution but I can't manage to shallow render components that use react-redux hooks.

@adngdb any hints? I'm really out of ideas.

**Comment by adngdb** [comment]
Hey @julen, I don't remember if it's possible to "shallow" components. There are however several examples of mounting components, like this one: https://github.com/mozilla/pontoon/blob/master/frontend/src/modules/genericeditor/components/GenericEditor.test.js#L44

Does that help, or are you looking for something else?

**Comment by julen** [comment]
Thanks @adngdb, that set me in the right track and I've been able to fix the tests... not without sweating though! I'd be grateful if you could now review the changes for me.

File: frontend/src/core/plural/components/PluralSelector.test.js [hunk 1]
```diff
@@ -78,20 +78,28 @@ describe('<PluralSelector>', () => {
             },
             router: {
                 location: {
-                    pathname: '/kg/pro/all/',
+                    pathname: '/kg/firefox/all-resources/',
                     search: '?string=42',
                 },
+                // `action` is required because
+                // https://github.com/supasate/connected-react-router/issues/312#issuecomment-500968504
```
Comment (julen) [inline frontend/src/core/plural/components/PluralSelector.test.js:85]: Super leaky implementation detail 🤦

File: frontend/src/core/plural/components/PluralSelector.test.js [hunk 2]
```diff
@@ -78,20 +78,28 @@ describe('<PluralSelector>', () => {
...
-            PluralSelectorBase,
-        );
+        const root = mountComponentWithStore(PluralSelector, store, {
+            resetEditor: sinon.spy(),
+        });
+        const wrapper = root.find(PluralSelectorBase);
 
-        wrapper.find('button').first().simulate('click');
+        const button = wrapper.find('button').first();
+        button.prop('onClick')();
```
Comment (julen) [inline frontend/src/core/plural/components/PluralSelector.test.js:102]: I really struggled with this one. Apparently `.simulate()` doesn't quite work when components are mounted (as opposed to shallow-rendered). If you read enzyme GitHub issues, the maintainer strongly suggests to stay away from `.simulate()` because it doesn't do what it appears to be doing, and instead suggests to directly invoke the component's handler function.

File: frontend/src/modules/entitieslist/components/EntitiesList.test.js [hunk 3]
```diff
@@ -11,7 +11,10 @@ import * as batchactions from 'modules/batchactions';
 import EntitiesList, { EntitiesListBase } from './EntitiesList';
 
 // Entities shared between tests
-const ENTITIES = [{ pk: 1 }, { pk: 2 }];
+const ENTITIES = [
+    { pk: 1, translation: [{ string: '', errors: [], warnings: [] }] },
```
Comment (julen) [inline frontend/src/modules/entitieslist/components/EntitiesList.test.js:15]: Because we are now deep rendering (i.e. `.mount()`), the passed in state needs to have more members so that children components don't crash and can render.

File: frontend/src/modules/entitydetails/components/EntityDetails.test.js [hunk 4]
```diff
@@ -21,6 +22,8 @@ const ENTITIES = [
                 warnings: [],
             },
         ],
+        project: { contact: '' },
```
Comment (julen) [inline frontend/src/modules/entitydetails/components/EntityDetails.test.js:25]: Same like before, this is to be able to deep render the component tree under test.

File: frontend/src/modules/entitydetails/components/EntityDetails.test.js [hunk 5]
```diff
@@ -209,9 +213,11 @@ describe('<EntityDetails>', () => {
     });
 
     it('dispatches the updateStatus action when updateTranslationStatus is called', () => {
-        const [wrapper] = createEntityDetailsWithStore();
+        const [wrapper, store] = createEntityDetailsWithStore();
 
         wrapper.instance().updateTranslationStatus(42, 'fake translation');
+        // Proceed with changes even if unsaved
+        store.dispatch(unsavedchanges.actions.ignore());
```
Comment (julen) [inline frontend/src/modules/entitydetails/components/EntityDetails.test.js:220]: I couldn't explain why this wasn't needed before, any ideas @adngdb?

File: frontend/src/test/store.js [hunk 6]
```diff
@@ -31,7 +31,7 @@ export function createReduxStore(initialState = {}) {
 export function mountComponentWithStore(Component, store, props = {}) {
     return mount(
         <Provider store={store}>
-            <ConnectedRouter history={history}>
+            <ConnectedRouter history={history} noInitialPop>
```
Comment (julen) [inline frontend/src/test/store.js:34]: This prop is required to omit an initial navigation dispatch from the router, which could have side-effects like resetting some initial state passed to the root reducer factory function. Please check the comment in _PluralSelector.test.js_ for further details.

File: frontend/src/modules/entitydetails/components/EntityDetails.test.js (hunk 5 above)
Comment (adngdb) [reply frontend/src/modules/entitydetails/components/EntityDetails.test.js:219]: That's weird. Maybe the unsavedchanges state is "polluted" and thus must be ignored before the action is actually called? You could check that by printing the value of `store.getState().unsavedchanges` before this line, to see if it has `exists` as true. As for why that would be, I do not know! Either a precedent test failed to clean the state, or there is something in the setup that triggers the unsavedchanges action.

File: frontend/src/test/store.js (hunk 6 above)
Comment (adngdb) [reply frontend/src/test/store.js:34]: Here again, can you write this down in the code too? Thanks!

**Review by adngdb — APPROVED** [review]
At first I was wary of the mix of class components and functional components, but I actually think this is a nice step towards rewriting everything to functions, so yay! I made a few comments, mostly about adding more comments. I like the graphs you showed, and I didn't find any bug in my quick testing, so great! :-)

File: frontend/src/modules/entitydetails/components/EntityDetails.test.js (hunk 5 above)
Comment (julen) [reply frontend/src/modules/entitydetails/components/EntityDetails.test.js:219]: Hmmm I went through the possible code paths again, and the main difference now being how `EntityDetails` is mounted, this also means underlying components are mounted, so the logic under `GenericTranslationForm` will be executed, which in turn fires the `unsavedchanges.actions.update()` action via the `useUpdateUnsavedChanges` hook.

And that brings us now to have the `unsavedchanges.actions.exists` populated to `true`, and because we are invoking `EntityDetail` component's `.updateTranslationStatus()` directly, that will only execute the `history.actions.updateStatus()` action when no unsaved changes exist (as part of the callback function).

Long story short: mounting brings us closer to real user interactions, and the line I added makes sense in this context.

File: frontend/src/core/plural/components/PluralSelector.test.js (hunk 2 above)
Comment (julen) [reply frontend/src/core/plural/components/PluralSelector.test.js:104]: Sure, done.

File: frontend/src/test/store.js (hunk 6 above)
Comment (julen) [reply frontend/src/test/store.js:37]: Done as well 👍

**Comment by julen** [comment]
> At first I was wary of the mix of class components and functional components, but I actually think this is a nice step towards rewriting everything to functions, so yay!

Yeah I'm not a big fan either (life was simpler when `React.createClass()` was the sole API!), but as I explained in the original PR comment, this was necessary in order to keep the scope of changes as small as possible. Initially I started to rewrite the components as functional components (as per the react-redux `useState()` hook requirement), but the scope of changes was going way too far away.

**Review by adngdb — APPROVED** [review]

**Comment by mathjazz** [comment]
Cool work, @julen! 👏 

Could you please squash commits? Same for #1833.

**Comment by julen** [comment]
Sure, done!
